staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation_test.go

```diff
@@ -92,3 +93,260 @@ func TestValidateLabels(t *testing.T) {
+			MatchLabels: map[string]string{
+				"id2": "app3",
+			},
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{Key: "foo", Operator: "Exists"},
+			},
+		},
+		{
+			MatchLabels: map[string]string{
+				"id1": "app2",
+			},
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{Key: "foo1", Operator: "DoesNotExist"},
+			},
+		},
+	}
+	for i := range successCases {
+		errs := ValidateLabelSelector(&successCases[i], field.NewPath("field"))
+		if len(errs) != 0 {
+			t.Errorf("successCases: case[%d]:%v expected success, got %#v", i, successCases[i], errs)
+		}
+	}
+
+	namePartErrMsg := "name part must consist of"
+	nameErrMsg := "a qualified name must consist of"
+	labelErrMsg := "a valid label must be an empty string or consist of"
+	maxLengthErrMsg := "must be no more than"
+
+	matchLabelErrorCases := []struct {
+		labelselector metav1.LabelSelector
+		expect        string
+	}{
+		{
+			labelselector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"ids^=@": "app2",
+					"id2":    "app3",
+				},
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{Key: "foo", Operator: "In", Values: []string{"bar1", "bar2"}},
+				},
+			},
+			expect: namePartErrMsg,
+		},
+		{
+			labelselector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"only/one/slas": "app2",
+					"id2":           "app3",
+				},
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{Key: "foo", Operator: "In", Values: []string{"bar1", "bar2"}},
+				},
+			},
+			expect: nameErrMsg,
+		},
+		{
+			labelselector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"only": "app2",
+					"good": "app3=@",
+				},
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{Key: "foo", Operator: "In", Values: []string{"bar1", "bar2"}},
+				},
+			},
+			expect: labelErrMsg,
+		},
+		{
+			labelselector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"only": "app2",
+					"good": strings.Repeat("a", 64),
+				},
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{Key: "foo", Operator: "In", Values: []string{"bar1", "bar2"}},
+				},
+			},
+			expect: maxLengthErrMsg,
+		},
+	}
+	for i := range matchLabelErrorCases {
+		errs := ValidateLabelSelector(&matchLabelErrorCases[i].labelselector, field.NewPath("field"))
+		if len(errs) == 0 {
+			t.Errorf("MatchLabelErrorCases: case[%d]:%v expected failure", i, matchLabelErrorCases[i])
+		} else {
+			if !strings.Contains(errs[0].Detail, matchLabelErrorCases[i].expect) {
+				t.Errorf("MatchLabelErrorCases: case[%d]:%v error details do not include %q: %q", i, matchLabelErrorCases[i], matchLabelErrorCases[i].expect, errs[0].Detail)
+			}
+		}
+	}
+
+	opInOrNotInErrMsg := "must be specified when `operator` is 'In' or 'NotIn'"
+	opExistsOrNotExistsErrMsg := "may not be specified when `operator` is 'Exists' or 'DoesNotExist'"
+	opNotValidErrMsg := "not a valid selector operator"
+
+	matchExpressionErrorCases := []struct {
```

Comment:
Why are these separate?  No real objection, just curious.

Reply:
when this unit test is failed, it helps to  which cases this failed.